packages/entity-server/src/routes/hierarchy/middleware/fields.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType, EntityFields } from '../../../models';
+import {
+  HierarchyRequest,
+  HierarchyContext,
+  ExtendedEntityFields,
+  EntityResponseObject,
+} from '../types';
+import {
+  extendedFieldFunctions,
+  getParentCodeBulk,
+  getChildrenCodesBulk,
+} from '../extendedFieldFunctions';
+
+const validFields: (keyof EntityFields)[] = ['code', 'country_code'];
+const isEntityField = (field: string): field is keyof EntityFields =>
+  (validFields as string[]).includes(field);
+const validateField = (field: string): field is keyof ExtendedEntityFields =>
+  isEntityField(field) || Object.keys(extendedFieldFunctions).includes(field);
+
+const allFields = [
+  ...validFields,
+  ...Object.keys(extendedFieldFunctions),
+] as (keyof ExtendedEntityFields)[];
+
+export const extractFieldsFromQuery = (queryFields?: string) => {
+  if (!queryFields) {
+    return allFields; // Display all fields if none specifically requested
+  }
+
+  const requestedFields = queryFields.split(',');
+  const fields = new Set<keyof ExtendedEntityFields>();
+  requestedFields.forEach(requestedField => {
+    if (validateField(requestedField)) {
+      fields.add(requestedField);
+    } else {
+      throw new Error(`Unknown field requested: ${requestedField}`);
+    }
+  });
+  return Array.from(fields);
+};
+
+type Writable<T> = { -readonly [field in keyof T]?: T[field] };
+export const mapEntityToFields = (fields: (keyof ExtendedEntityFields)[]) => async (
+  entity: EntityType,
+  context: HierarchyContext,
+) => {
+  const mappedEntity: Writable<EntityResponseObject> = {};
+  for (let i = 0; i < fields.length; i++) {
+    const field = fields[i];
+    if (isEntityField(field)) {
+      mappedEntity[field] = entity[field];
+    } else {
+      (mappedEntity[field] as string | string[] | undefined) = await extendedFieldFunctions[field](
+        entity,
+        context,
+      );
+    }
+  }
+  return mappedEntity;
+};
+
+export const mapEntitiesToFields = (
```

Comment:
This function is right on the borderline for me of unreadability (switch statement within an if/else within a loop...) but I'm okay with how it is for now as atleast it encapsulates a lot of related funky logic in the one place. Welcome for pushback on it though

Reply:
I'm actually fine with it - it reads pretty clearly to me tbh (though I do have the context of the rest of the PR in my head)